components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/SecureAbove22Preferences.kt

```diff
@@ -0,0 +1,168 @@
+
+    private val logger = Logger("SecureAbove22Preferences")
+    private val prefs = if (Build.VERSION.SDK_INT < M) {
+        context.getSharedPreferences(KEY_PREFERENCES_PRE_M, MODE_PRIVATE)
+    } else {
+        context.getSharedPreferences(KEY_PREFERENCES_POST_M, MODE_PRIVATE)
+    }
+
+    init {
+        // If we're running on an API level for which we support encryption, see if we have any plaintext values stored
+        // on disk. That indicates that we've hit an API upgrade situation - we just went from pre-M to post-M. Since
+        // we already have the plaintext keys, we can transparently migrate them to use the encrypted storage layer.
+        if (Build.VERSION.SDK_INT >= M) {
+            val plaintextPrefs = context.getSharedPreferences(KEY_PREFERENCES_PRE_M, MODE_PRIVATE)
+            if (plaintextPrefs.all.isNotEmpty()) {
+                migratePrefs(plaintextPrefs)
+            }
+        }
+    }
+
+    /**
+     * Retrieves a stored [key]. See [putString] for storing a [key].
+     *
+     * @param key A key name.
+     * @return An optional [String] if [key] is present in the store.
+     */
+    fun getString(key: String): String? {
+        return if (Build.VERSION.SDK_INT >= M) {
+            getStringDecrypted(key)
+        } else {
+            prefs.getString(key, null)
+        }
+    }
+
+    /**
+     * Stores [value] under [key]. Retrieve it using [getString].
+     *
+     * @param key A key name.
+     * @param value A value for [key].
+     */
+    fun putString(key: String, value: String) {
+        if (Build.VERSION.SDK_INT >= M) {
+            putStringEncrypted(key, value)
+        } else {
+            prefs.edit().putString(key, value).apply()
+        }
+    }
+
+    /**
+     * Removes key/value pair from storage for the provided [key].
+     */
+    fun remove(key: String) {
+        prefs.edit().remove(key).apply()
+    }
+
+    @TargetApi(M)
+    private fun getStringDecrypted(key: String): String? {
+        // The fact that we're possibly generating a managed key here implies that this key could be lost after being
+        // for some reason. One possible reason for a key to be lost is rotating signing keys for the APK.
+        // Applications are encouraged to instrument such events.
+        generateManagedKeyIfNecessary()
+
+        return if (prefs.contains(key)) {
+            val value = prefs.getString(key, "")
+            val encrypted = Base64.decode(value, BASE_64_FLAGS)
+
+            val plain = try {
+                keystore!!.decryptBytes(encrypted)
+            } catch (error: IllegalArgumentException) {
+                logger.error("IllegalArgumentException exception: ", error)
+                null
+            } catch (error: GeneralSecurityException) {
+                logger.error("Decrypt exception: ", error)
+                null
+            }
+
+            plain.let {
+                if (it != null) String(it, StandardCharsets.UTF_8) else null
+            }
```

Comment:
Kotlin nit: For one I think `let` makes code pretty hard to read if you are just doing a null check on an immutable(!!) variable (that's just a fancy "if"). And here it's even more weird since it doesn't do more than translating `plain` to `it`.

There's no difference to this (since `plain` is immutable!):

```Kotlin
 if (plain != null) String(plain, StandardCharsets.UTF_8) else null
```

Since let also returns a value you can use it for the nullability check here though, especially since you want `null` in the else case (no need to move the null check inside):

```Kotlin
plain?.let { String(it, StandardCharsets.UTF_8) }   // this statement will already be null in the "else" case.
```

But I think a good old `if` is just more readable. `let` only seems to be helpful if you are dealing with mutable variables that are not in your control.